Add some UI improvements to Value Tracking

ryzngard opened the pull request:
* Navigate to items on selection
![value_tracking_navigate](https://user-images.githubusercontent.com/475144/113919530-8eb54680-9798-11eb-98b2-729694e794d4.gif)


* Highlight the span that is relevant for expansion

![image](https://user-images.githubusercontent.com/475144/113919338-56156d00-9798-11eb-812b-78e43f5b0b11.png)

JoeRobich commented on src/EditorFeatures/Core/ValueTracking/ValueTrackingService.Visitor.cs:
```diff
@@ -249,7 +249,7 @@ or IBinaryOperation
                     return true;
                 }
-
+                
```
Remove indentation.
```suggestion

```

JoeRobich commented on src/EditorFeatures/Test/ValueTracking/AbstractBaseValueTrackingTests.cs:
```diff
@@ -74,8 +75,9 @@ internal static async Task ValidateChildrenEmptyAsync(TestWorkspace testWorkspac
         internal static void ValidateItem(ValueTrackedItem item, int line, string? text = null)
-        {
-            Assert.Equal(line, item.LineSpan.Start);
+{
```
Add indentation.
```suggestion
        {
```

JoeRobich reviewed (approved):
Looks good other than some indentation issues.

CyrusNajmabadi commented on src/EditorFeatures/Core/ValueTracking/ValueTrackingService.Visitor.cs:
```diff
@@ -249,7 +249,6 @@ or IBinaryOperation
                 {
                     return true;
                 }
-
```
revert plz :)

CyrusNajmabadi commented on src/VisualStudio/Core/Def/ValueTracking/ValueTrackingCommandHandler.cs:
```diff
@@ -144,6 +143,12 @@ private async Task ShowToolWindowAsync(ITextView textView, ISymbol selectedSymbo
             }
 
+            var toolWindow = await GetOrCreateToolWindowAsync(textView, cancellationToken).ConfigureAwait(false);
+            if (toolWindow is null || toolWindow.ViewModel is null)
```
can this happen?

ryzngard replied:
It shouldn't, but at most we should log a problem and not crash here imo

CyrusNajmabadi replied:
my request is that we either log, or doc this code.  right now it looks like we're just silent.  which makes me feel like a customer will try to use this feature, see nothing appear and have no idea what's going on :)

CyrusNajmabadi commented on src/VisualStudio/Core/Def/ValueTracking/ValueTrackingCommandHandler.cs:
```diff
@@ -144,6 +143,12 @@ private async Task ShowToolWindowAsync(ITextView textView, ISymbol selectedSymbo
             }
 
+            var toolWindow = await GetOrCreateToolWindowAsync(textView, cancellationToken).ConfigureAwait(false);
+            if (toolWindow is null || toolWindow.ViewModel is null)
```
```suggestion
            if (toolWindow?.ViewModel is null)
```

CyrusNajmabadi commented on src/VisualStudio/Core/Def/ValueTracking/ValueTrackingCommandHandler.cs:
```diff
@@ -154,43 +159,57 @@ private async Task ShowToolWindowAsync(ITextView textView, ISymbol selectedSymbo
+            Contract.ThrowIfNull(roslynPackage);
+
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+            await roslynPackage.ShowToolWindowAsync(
```
it's a pity ShowToolWindow doesn't do teh switch, given that's async.  @sharwell should we invert this?

ryzngard replied:
It may, but I haven't found good documentation on this. AFAIK switchtomain should no-op if it's already on main so the work is done regardless of who does it

jmarolf replied:
[internal link](https://devdiv.visualstudio.com/DevDiv/_git/VS?path=%2Fsrc%2Fenv%2Fshell%2FPackageFramework%2FCurrent%2FShell%2FAsyncPackage.cs&version=GBmain&line=1045&lineEnd=1046&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents) looks like this already switched to the main thread internally.

CyrusNajmabadi commented on src/VisualStudio/Core/Def/ValueTracking/ValueTrackingToolWindow.cs:
```diff
@@ -13,29 +13,33 @@ namespace Microsoft.VisualStudio.LanguageServices.ValueTracking
 
-        public ValueTrackingToolWindow(ValueTrackingTreeItemViewModel root)
-            : base(null)
+        public ValueTrackingToolWindow() : base(null)
```
what is this contructor for?

ryzngard replied:
If a user has the tool window open already, this is the constructor that gets called

CyrusNajmabadi commented on src/VisualStudio/Core/Def/ValueTracking/ValueTrackingTreeItemViewModel.cs:
```diff
@@ -16,20 +16,31 @@
+using Microsoft.CodeAnalysis.Navigation;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
```
sort

CyrusNajmabadi commented on src/VisualStudio/Core/Def/ValueTracking/ValueTrackingTreeItemViewModel.cs:
```diff
@@ -42,34 +53,55 @@ public IList<Inline> Inlines
+                    TreeViewModel.ClassificationTypeMap,
+                    (run, classifiedText, position) =>
+                    {
+                        if (TreeViewModel.HighlightBrush is not null)
```
when would tihs be null?

ryzngard replied:
If it's not set in the properties, same as https://github.com/dotnet/roslyn/blob/main/src/VisualStudio/Core/Def/Implementation/FindReferences/Entries/DocumentSpanEntry.cs#L94

CyrusNajmabadi commented:
looks awesome. a general complaint i have is that a lot of the code seems to have null handling that seems odd to me.  i can't tell if it's just being paranoid, or because these values truly can be null.  the latter seems odd to me as i don't really see how this scenario should ever end up with nulls/failures in these cases

jmarolf reviewed (approved):
:shipit:

jmarolf commented:
@ryzngard I assume label that are needed for accessibility can be added later lets just remember to update the xaml before we ship

ryzngard commented:
> @ryzngard I assume label that are needed for accessibility can be added later lets just remember to update the xaml before we ship

Correct. No accessibility pass has been done at this point

jmarolf commented:
@ryzngard one comment I have is that we may want to "summarize" the line if it gets too long. I can't even see what is on the right side in you gif.

perhaps something like this? `S = ... + this.GetString() + ... [MyClass.cs(18, 88)]`

mikadumont reviewed (changes requested):
I cant tell if the example is confusing but its hard to tell what the value is. Can we bold it instead of highlight it in grey? Also do we need the whole line of information since we know from the above node and can possibly shorten it to show the value; `S = **this.GetString()** [MyClass.cs (line number)]

ryzngard commented:
Updated to this

![image](https://user-images.githubusercontent.com/475144/114474439-1640f200-9bab-11eb-8db5-41b7bd2e01ab.png)


cc @mikadumont  @jmarolf

jmarolf commented:
@ryzngard why do we show the entire line for the third item?

CyrusNajmabadi commented:
honestly, teh truncation is super confusing to me.  i liked the original version where it was the line, but the relevant section had the gray background.

ryzngard commented:
> @ryzngard why do we show the entire line for the third item?

That's a good question, we shouldn't be... I also need to add some tests for this

CyrusNajmabadi commented:
![image](https://user-images.githubusercontent.com/4564579/114475584-6a4cd600-9bad-11eb-8046-18ba32946090.png)

This was super clear and easy to grok.  Also looks like FAR to me, which is nice.

jmarolf commented:
I would like it to still have the grey background and just go to `...` when it get more than 80 characters long (which should be rare hopefully)

CyrusNajmabadi commented:
> I would like it to still have the grey background and just go to ... when it get more than 80 characters long (which should be rare hopefully)

I'm really not a fan.  We don't do that with FAR.  And it seems like this woudl naturally just fall out from having columns in the UI.

ryzngard commented:
> I'm really not a fan. We don't do that with FAR. And it seems like this woudl naturally just fall out from having columns in the UI.

I'm not sure how columns would help in this case, the only information that would be pushed out would be the file name and line right?

ryzngard commented:
> I would like it to still have the grey background and just go to ... when it get more than 80 characters long (which should be rare hopefully)

@mikadumont  was pretty adamant about doing the bold instead of highlight. I'm indifferent, and both are easy enough to do.

CyrusNajmabadi commented:
fwiw, the scattershot `[MyClass:17]` bits everywhere are not helping.  I would def like tehse to be columns.  One foe 'containing type' (or file) and one for 'line'.  It feels super weird to me to have a tree control with lots of pieces of data, and to pull so far away from FAR.  i get that we're not using that particular control.  But i still feel it should behave sensibly similar.

CyrusNajmabadi commented:
> @mikadumont was pretty adamant about doing the bold instead of highlight. I'm indifferent, and both are easy enough to do.

I prefer the gray, but the bold seems ok (though it might be hard to see in some cases, i'm not srue).  What i def do not like is the truncation.  Indeed, i *really* want to see what's going around this chunk so i can understand the context around this easily just from that UI.

Hovering can also be used to get a few lines of context above/below, like so:

![image](https://user-images.githubusercontent.com/4564579/114475946-2b6b5000-9bae-11eb-9843-d941568a3357.png)

ryzngard commented:
>  I would def like tehse to be columns

I still strongly believe columns would not work for deep nested trees. We could add a "details" pane like Call Hierarchy has, but I don't think a column setup would work well for data where indentation is part of the data

CyrusNajmabadi commented:
> don't think a column setup would work well for data where indentation is part of the data

I mean, we have lots of UIs that do this.  Things like perf view and whatnot.  I think it's fine to do this.  There's just a need (*regardless of columns*) to be able to say: i need to dedent and treat this as the root.

Consider just a long chain that i have to dive into.  I'm on a 14" laptop screen most of the time.  I'm going to quickly blow past the space i have as i open this up.  If, however, i can take nodes i care about and make them roots, then that's not an issue.  It also helps out if there are a few different paths i need to go down.

--

Finally, if you aren't going to do columns, can we just make it so that the file/line is right aligned, instead of being left aligned.  It just looks so chaotic to me to try to decipher:

![image](https://user-images.githubusercontent.com/4564579/114476814-d03a5d00-9baf-11eb-8a2f-78bffdab66e0.png)

CyrusNajmabadi commented:
Without columns, but right aligned:

![image](https://user-images.githubusercontent.com/4564579/114476923-198aac80-9bb0-11eb-8b60-179cad08a589.png)

I think that's *much* cleaner and easier to parse.  Though i still question why we're showing things as `[TypeName:#]` as taht's just not a metaphor we have anywhere else afaict.

mikadumont commented:
> Without columns, but right aligned:
> 
> ![image](https://user-images.githubusercontent.com/4564579/114476923-198aac80-9bb0-11eb-8b60-179cad08a589.png)
> 
> I think that's _much_ cleaner and easier to parse. Though i still question why we're showing things as `[TypeName:#]` as taht's just not a metaphor we have anywhere else afaict.

I believe its [FileName: line#] 

@ryzngard I do think this line is the most clear though and the truncation isn't clear tbh:
![image](https://user-images.githubusercontent.com/46729679/114605167-a6844300-9c4e-11eb-99fb-e8bd7ba162d8.png)

Possibly because `S =` is at the beginning of the line and the `...` seemed to abruptly starts at a strange cut off point.

Between the bold and grey highlight I don't think the bold stands out as much as the grey highlight and think we should switch it back.

ryzngard commented:
> @ryzngard I do think this line is the most clear though and the truncation isn't clear tbh 

That's a bug. It should be truncated. 

Also note the truncation is really aggressive right now at 20 characters. 120 would not truncate this line at all

ryzngard commented:
Updated to remove bold and go back to highlight, similar to FAR. I've removed truncation for now, let's use #52590 to track if we want it. 

I'm going to add an extra panel for details and remove [file name:line] at the end for clarity.

ryzngard commented:
Updated to this. Left and right are columns, gridsplitter inbetween to resize ![image](https://user-images.githubusercontent.com/475144/114625828-af354300-9c67-11eb-9dba-e0bce0b552a9.png)
